src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs

```diff
@@ -328,125 +284,317 @@ internal static string ReadString(ref Utf8JsonReader reader, string propertyName
+            else if (obj is bool b)
+                writer.WriteBoolean(key, b);
+            else if (obj is decimal d)
+                writer.WriteNumber(key, d);
+            else if (obj is double dub)
+                writer.WriteNumber(key, dub);
+            else if (obj is float f)
+                writer.WriteNumber(key, f);
+            else if (obj is long l)
+                writer.WriteNumber(key, l);
             else if (obj is null)
                 writer.WriteNull(key);
-            else if (obj is JsonElement)
+            else if (obj is List<string> strs)
+            {
+                writer.WriteStartArray(key);
+                foreach (string item in strs)
+                    writer.WriteStringValue(item);
+
+                writer.WriteEndArray();
+            }
+            else if (depth < MaxDepth && obj is List<object> objs)
+            {
+                depth++;
+                writer.WriteStartArray(key);
+                foreach (object item in objs)
+                    WriteObjectValue(ref writer, item, depth);
+
+                writer.WriteEndArray();
+            }
+            else if (obj is IDictionary<string, string> idics)
+            {
+                writer.WriteStartObject(key);
+                foreach (KeyValuePair<string, string> kvp in idics)
+                    writer.WriteString(kvp.Key, kvp.Value);
+
+                writer.WriteEndObject();
+            }
+            else if (depth < MaxDepth && obj is IDictionary<string, object> idic)
+            {
+                depth++;
+                writer.WriteStartObject(key);
+                foreach (KeyValuePair<string, object> kvp in idic)
+                    WriteObject(ref writer, kvp.Key, kvp.Value, depth);
+
+                writer.WriteEndObject();
+            }
+            else if (depth < MaxDepth && obj is Dictionary<string, object> dic)
+            {
+                depth++;
+                writer.WriteStartObject(key);
+                foreach (KeyValuePair<string, object> kvp in dic)
+                    WriteObject(ref writer, kvp.Key, kvp.Value, depth);
+
+                writer.WriteEndObject();
+            }
+            else if (obj is JsonElement j)
             {
                 writer.WritePropertyName(key);
-                ((JsonElement)obj).WriteTo(writer);
+                j.WriteTo(writer);
             }
             else
             {
                 writer.WriteString(key, obj.ToString());
             }
         }
 
-        internal static void WriteStrings(ref Utf8JsonWriter writer, ReadOnlySpan<byte> propertyName, IList<string> strings)
+        /// <summary>
+        /// Writes values into an array.
+        /// Assumes the writer.StartArray() has been called.
+        /// </summary>
+        /// <param name="writer"></param>
+        /// <param name="obj"></param>
+        /// <param name="depth">The current depth of recursive call for objects.
+        /// Maximum is 3.</param>
```

Comment:
isn't max depth 2?

Reply:
@keegan-caruso , yea i kept going back and forth.
fixed.